File: model-client/src/main/java/org/modelix/model/persistent/CPVersion.java
package org.modelix.model.persistent;

import org.apache.commons.lang3.StringUtils;
import org.apache.log4j.Level;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.modelix.model.operations.IOperation;

import java.util.stream.Stream;

public class CPVersion {
  private static final Logger LOG = LogManager.getLogger(CPVersion.class);

  public final long id;
  public final String time;
  public final String author;

  /**
   * SHA to CPTree
   */
  public final String treeHash;
  public final String previousVersion;
  public final IOperation[] operations;
  public final String operationsHash;
  public final int numberOfOperations;

  public CPVersion(long id, String time, String author, String treeHash, String previousVersion, IOperation[] operations, String operationsHash, int numberOfOperations) {
    if ((treeHash == null || treeHash.length() == 0)) {
      if (LOG.isEnabledFor(Level.WARN)) {
        LOG.warn("No tree hash provided", new Exception());
      }
    }
    if ((operations == null) == (operationsHash == null)) {
      throw new RuntimeException("Only one of 'operations' and 'operationsHash' can be provided");
    }
    this.id = id;
    this.author = author;
    this.previousVersion = previousVersion;
    this.time = time;
    this.treeHash = treeHash;
    this.operations = operations;
    this.operationsHash = operationsHash;
    this.numberOfOperations = (operations != null ? operations.length : numberOfOperations);
  }

  public String serialize() {
    String opsPart = operationsHash != null
            ? operationsHash :
            Stream.of(operations)
                    .map(OperationSerializer.INSTANCE::serialize)
                    .reduce((a, b) -> a + "," + b)
                    .orElse("");
    String serialized = SerializationUtil.longToHex(id) +
            "/" + SerializationUtil.escape(time) +
            "/" + SerializationUtil.escape(author) +
            "/" + SerializationUtil.nullAsEmptyString(treeHash) +
            "/" + SerializationUtil.nullAsEmptyString(previousVersion) +
            "/" + opsPart;
    if (numberOfOperations >= 0) {
      serialized += "/" + numberOfOperations;
    }
    return serialized;
  }

  public static CPVersion deserialize(String input) {
    String[] parts = input.split("/", -1);

    String opsHash = null;
    IOperation[] ops = null;
    if (HashUtil.isSha256(parts[5])) {
      opsHash = parts[5];
    } else {
      ops = Stream.of(parts[5].split(","))
              .filter(StringUtils::isNotEmpty)
              .map(OperationSerializer.INSTANCE::deserialize)
              .toArray(IOperation[]::new);
    }

    int numOps = (parts.length >= 7 ? Integer.parseInt(parts[6]) : -1);
    return new CPVersion(
            SerializationUtil.longFromHex(parts[0]),
            SerializationUtil.unescape(parts[1]),
            SerializationUtil.unescape(parts[2]),
            SerializationUtil.emptyStringAsNull(parts[3]),
            SerializationUtil.emptyStringAsNull(parts[4]),
            ops,
            opsHash,
            numOps);
  }

  public String getHash() {
    return HashUtil.sha256(this.serialize());
  }

}

File: model-client/src/main/kotlin/org/modelix/model/persistent/CPVersion.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.modelix.model.persistent

import org.apache.commons.lang3.StringUtils
import org.apache.log4j.Level
import org.apache.log4j.LogManager
import org.modelix.model.operations.IOperation
import org.modelix.model.persistent.HashUtil.isSha256
import org.modelix.model.persistent.HashUtil.sha256
import org.modelix.model.persistent.SerializationUtil.emptyStringAsNull
import org.modelix.model.persistent.SerializationUtil.escape
import org.modelix.model.persistent.SerializationUtil.longFromHex
import org.modelix.model.persistent.SerializationUtil.longToHex
import org.modelix.model.persistent.SerializationUtil.nullAsEmptyString
import org.modelix.model.persistent.SerializationUtil.unescape
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.Stream

class CPVersion(id: Long, time: String?, author: String?, treeHash: String?, previousVersion: String?, operations: Array<IOperation>?, operationsHash: String?, numberOfOperations: Int) {
    val id: Long
    val time: String?
    val author: String?

    /**
     * SHA to CPTree
     */
    val treeHash: String?
    val previousVersion: String?
    val operations: Array<IOperation>?
    val operationsHash: String?
    val numberOfOperations: Int
    fun serialize(): String {
        val opsPart = operationsHash
            ?: operations!!.toList().stream()
                .map(Function<IOperation?, String> { op: IOperation? -> OperationSerializer.INSTANCE.serialize(op!!) })
                .reduce { a: String, b: String -> "$a,$b" }
                .orElse("")
        var serialized = longToHex(id) +
            "/" + escape(time) +
            "/" + escape(author) +
            "/" + nullAsEmptyString(treeHash) +
            "/" + nullAsEmptyString(previousVersion) +
            "/" + opsPart
        if (numberOfOperations >= 0) {
            serialized += "/$numberOfOperations"
        }
        return serialized
    }

    val hash: String
        get() = sha256(serialize())

    companion object {
        private val LOG = LogManager.getLogger(CPVersion::class.java)
        fun deserialize(input: String): CPVersion {
            val parts = input.split("/").dropLastWhile { it.isEmpty() }.toTypedArray()
            var opsHash: String? = null
            var ops: Array<IOperation>? = null
            if (isSha256(parts[5])) {
                opsHash = parts[5]
            } else {
                ops = Stream.of(*parts[5].split(",").toTypedArray())
                    .filter { cs: String? -> StringUtils.isNotEmpty(cs) }
                    .map { serialized: String? -> OperationSerializer.INSTANCE.deserialize(serialized!!) }
                    .collect(Collectors.toList()).toTypedArray()
            }
            val numOps = if (parts.size >= 7) parts[6].toInt() else -1
            return CPVersion(
                longFromHex(parts[0]),
                unescape(parts[1]),
                unescape(parts[2]),
                emptyStringAsNull(parts[3]),
                emptyStringAsNull(parts[4]),
                ops,
                opsHash,
                numOps
            )
        }
    }

    init {
        if (treeHash == null || treeHash.length == 0) {
            if (LOG.isEnabledFor(Level.WARN)) {
                LOG.warn("No tree hash provided", Exception())
            }
        }
        if (operations == null == (operationsHash == null)) {
            throw RuntimeException("Only one of 'operations' and 'operationsHash' can be provided")
        }
        this.id = id
        this.author = author
        this.previousVersion = previousVersion
        this.time = time
        this.treeHash = treeHash
        this.operations = operations
        this.operationsHash = operationsHash
        this.numberOfOperations = operations?.size ?: numberOfOperations
    }
}
